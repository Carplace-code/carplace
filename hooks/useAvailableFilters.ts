// hooks/useAvailableFilters.ts
import type { Prisma } from "@prisma/client";
import { useQuery } from "@tanstack/react-query";
import axios from "axios";

export function useAvailableFilters(where: Prisma.VersionWhereInput) {
  return useQuery({
    queryKey: ["availableFilters", where],
    queryFn: async () => {
      const res = await axios.post("/api/filters", { where });
      return res.data as {
        brands: string[];
        models: string[];
        years: number[];
        bodyTypes: string[];
      };
    },
    staleTime: 60 * 1000, // cache por 1 min
  });
}
