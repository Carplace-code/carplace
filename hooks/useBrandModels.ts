import { useQuery } from "@tanstack/react-query";

import axios from "@/lib/axios";

export function useBrandModels() {
  return useQuery<Record<string, string[]>>({
    queryKey: ["brandModels"],
    queryFn: () => axios.get("/brand-models").then((res) => res.data),
  });
}
