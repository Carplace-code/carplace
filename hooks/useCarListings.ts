// src/hooks/useListings.ts
import { Prisma } from "@prisma/client";
import { useQuery } from "@tanstack/react-query";

import axios from "@/lib/axios";

type CarListingWithInclude<TInclude extends Prisma.CarListingInclude> = Prisma.CarListingGetPayload<{
  include: TInclude;
}>;

interface UseGetCarListingsOptions<
  TInclude extends Prisma.CarListingInclude,
  TOrderBy extends Prisma.CarListingOrderByWithRelationInput,
> {
  include?: TInclude;
  pageSize?: number;
  orderBy?: TOrderBy;
}

export function useGetCarListings<
  TInclude extends Prisma.CarListingInclude,
  TOrderBy extends Prisma.CarListingOrderByWithRelationInput,
>(options?: UseGetCarListingsOptions<TInclude, TOrderBy>) {
  const { include, pageSize, orderBy } = options || {};

  return useQuery<CarListingWithInclude<TInclude>[], Error>({
    queryKey: ["car_listings", include, pageSize, orderBy],
    queryFn: () =>
      axios
        .get<CarListingWithInclude<TInclude>[]>("/car_listings", {
          params: {
            include: include ? JSON.stringify(include) : undefined,
            pageSize,
            orderBy: orderBy ? JSON.stringify(orderBy) : undefined,
          },
        })
        .then((res) => res.data),
  });
}

export function useGetCarListing<TInclude extends Prisma.CarListingInclude = object>({
  id,
  include,
}: {
  id?: string;
  include?: TInclude;
}) {
  return useQuery<CarListingWithInclude<TInclude>, Error>({
    queryKey: ["car_listing", id, include],
    enabled: Boolean(id),
    queryFn: () =>
      axios
        .get<CarListingWithInclude<TInclude>>(`/car_listings/${id}`, {
          params: { include: include ? JSON.stringify(include) : undefined },
        })
        .then((res) => res.data),
  });
}
