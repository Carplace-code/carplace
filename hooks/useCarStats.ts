"use client";

import { useEffect, useState } from "react";

import { getCarListingsStats } from "@/lib/actions/stats";

interface CarStats {
  totalListings: number;
  totalBrands: number;
  recentListings: number;
  success: boolean;
}

export function useCarStats() {
  const [stats, setStats] = useState<CarStats>({
    totalListings: 0,
    totalBrands: 0,
    recentListings: 0,
    success: false,
  });
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    async function fetchStats() {
      try {
        const result = await getCarListingsStats();
        setStats(result);
      } catch (error) {
        setStats({
          totalListings: 0,
          totalBrands: 0,
          recentListings: 0,
          success: false,
        });
      } finally {
        setIsLoading(false);
      }
    }

    fetchStats();
  }, []);

  return { stats, isLoading };
}
