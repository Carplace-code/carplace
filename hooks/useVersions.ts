// src/hooks/useVersions.ts
import { Prisma } from "@prisma/client";
import { useQuery } from "@tanstack/react-query";

import axios from "@/lib/axios";

type VersionWithInclude<TInclude extends Prisma.VersionInclude> = Prisma.VersionGetPayload<{ include: TInclude }>;

interface Meta {
  total: number;
  page: number;
  pageSize: number;
  pageCount: number;
}

/** All the parameters your API supports */
interface UseGetVersionsOptions<
  TInclude extends Prisma.VersionInclude,
  TWhere extends Prisma.VersionWhereInput,
  TOrderBy extends Prisma.VersionOrderByWithRelationInput,
> {
  include?: TInclude;
  where?: TWhere;
  orderBy?: TOrderBy;
  page?: number;
  pageSize?: number;
}

export function useGetVersions<
  TInclude extends Prisma.VersionInclude,
  TWhere extends Prisma.VersionWhereInput,
  TOrderBy extends Prisma.VersionOrderByWithRelationInput,
>(options?: UseGetVersionsOptions<TInclude, TWhere, TOrderBy>) {
  const { include, where, orderBy, page, pageSize } = options || {};

  return useQuery<{ data: VersionWithInclude<TInclude>[]; meta: Meta }, Error>({
    queryKey: ["versions", include, where, orderBy, page, pageSize],
    queryFn: () =>
      axios
        .get<{ data: VersionWithInclude<TInclude>[]; meta: Meta }>("/versions", {
          params: {
            include: include ? JSON.stringify(include) : undefined,
            where: where ? JSON.stringify(where) : undefined,
            orderBy: orderBy ? JSON.stringify(orderBy) : undefined,
            page,
            pageSize,
          },
        })
        .then((res) => res.data),
  });
}
export function useGetVersion<TInclude extends Prisma.VersionInclude = object>({
  id,
  include,
}: {
  id: string;
  include?: TInclude;
}) {
  return useQuery<VersionWithInclude<TInclude>, Error>({
    queryKey: ["version", id, include],
    enabled: Boolean(id),
    queryFn: () =>
      axios
        .get<VersionWithInclude<TInclude>>(`/versions/${id}`, {
          params: {
            include: include ? JSON.stringify(include) : undefined,
          },
        })
        .then((res) => res.data),
  });
}
