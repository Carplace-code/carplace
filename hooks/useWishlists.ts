import { Prisma, Wishlist } from "@prisma/client";
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import axios from "@/lib/axios";

type WishlistWithInclude<TInclude extends Prisma.WishlistInclude> = Prisma.WishlistGetPayload<{
  include: TInclude;
}>;

interface UseGetWishlistOptions<TInclude extends Prisma.WishlistInclude> {
  include?: TInclude;
}

export function useGetWishlists<TInclude extends Prisma.WishlistInclude>(
  { versionId, enabled, include }: { versionId?: string; enabled?: boolean } & UseGetWishlistOptions<TInclude> = {
    enabled: true,
  },
) {
  return useQuery<WishlistWithInclude<TInclude>[], Error>({
    queryKey: ["wishlists", versionId, include],
    queryFn: () =>
      axios
        .get("/wishlists", { params: { versionId, include: include ? JSON.stringify(include) : undefined } })
        .then((res) => res.data),
    enabled,
  });
}

export function useGetWishlist({ wishlistId }: { wishlistId: string }) {
  return useQuery<Wishlist, Error>({
    queryKey: ["wishlist", wishlistId],
    queryFn: () => axios.get(`/wishlists/${wishlistId}`).then((res) => res.data),
  });
}

export function useCreateWishlist() {
  const qc = useQueryClient();

  return useMutation<Wishlist, Error, Pick<Wishlist, "name" | "description">>({
    mutationFn: (newWishlist) => axios.post<Wishlist>("/wishlists", newWishlist).then((res) => res.data),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["wishlists"] });
    },
    onError: () => {},
  });
}
export function useUpdateWishlist() {
  const qc = useQueryClient();

  return useMutation<Wishlist, Error, Pick<Wishlist, "id"> & Partial<Pick<Wishlist, "name" | "description">>>({
    mutationFn: (wishlistUpdates) =>
      axios.patch(`/wishlists/${wishlistUpdates.id}`, wishlistUpdates).then((res) => res.data),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["wishlists"] });
    },
    onError: () => {},
  });
}

export function useDeleteWishlist() {
  const qc = useQueryClient();

  return useMutation<void, Error, string>({
    mutationFn: (wishlistId) => axios.delete(`/wishlists/${wishlistId}`),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["wishlists"] });
    },
    onError: () => {},
  });
}

export function useAddToWishlist() {
  const qc = useQueryClient();

  return useMutation<Wishlist, Error, { wishlistId: string; versionId: string }>({
    mutationFn: ({ wishlistId, versionId }) =>
      axios.post(`/wishlists/${wishlistId}/versions`, { versionId }).then((res) => res.data),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["wishlists"] });
    },
    onError: () => {},
  });
}

export function useRemoveFromWishlist() {
  const qc = useQueryClient();

  return useMutation<void, Error, { wishlistId: string; versionId: string }>({
    mutationFn: ({ wishlistId, versionId }) => axios.delete(`/wishlists/${wishlistId}/versions/${versionId}`),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["wishlists"] });
    },
    onError: () => {},
  });
}
