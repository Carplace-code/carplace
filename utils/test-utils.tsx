// tests/test-utils.tsx
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { render } from "@testing-library/react";
import { ReactNode } from "react";

// Puedes extender esta función si necesitas envolver en más contextos
export function renderWithQueryClient(children: ReactNode) {
  const queryClient = new QueryClient();
  return render(<QueryClientProvider client={queryClient}>{children}</QueryClientProvider>);
}
