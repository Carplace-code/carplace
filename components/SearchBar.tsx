/* eslint-disable */
"use client";

import { Input } from "@/components/ui/input";
import { useGetVersions } from "@/hooks/useVersions";
import type { Prisma } from "@prisma/client";
import { Search } from "lucide-react";
import { useRouter } from "next/navigation";
import type React from "react";
import { useEffect, useMemo, useRef, useState } from "react";

type VersionWithModelBrand = Prisma.VersionGetPayload<{
  include: {
    model: {
      include: {
        brand: true
      }
    }
  }
}>;

export default function SearchBar() {
  const [searchQuery, setSearchQuery] = useState("");
  const [filteredVersions, setFilteredVersions] = useState<VersionWithModelBrand[]>([]);
  const [showResults, setShowResults] = useState(false);
  const searchRef = useRef<HTMLDivElement>(null);
  const inputRef = useRef<HTMLInputElement>(null);

  const where = useMemo<Prisma.VersionWhereInput>(() => ({
    trims: {
      some: {
        carListings: {
          some: {},
        },
      },
    },
  }), []);

  const { data } = useGetVersions({
    where,
    include: {
      model: {
        include: {
          brand: true,
        },
      },
    },
    pageSize: 5000, // Fetch a larger set to allow for better search results
  });

  const versions = data?.data ?? [];

  useEffect(() => {
  if (!data?.data || !Array.isArray(versions)) return;

  if (!searchQuery.trim()) {
    setFilteredVersions([]);
    setShowResults(false);
    return;
  }

  const searchTerm = searchQuery.toLowerCase();

  const filtered = versions.filter((version) => {;
    const brand = version.model?.brand?.name ?? "";
    const model = version.model?.name ?? "";
    const year = version.year.toString();
    const fullLabel = `${brand} ${model} ${year}`.toLowerCase();
    return fullLabel.includes(searchTerm);
  });

  setFilteredVersions(filtered.slice(0, 8));
  setShowResults(true);
}, [searchQuery, versions]);

  const router = useRouter();

  const handleVersionSelect = (version: VersionWithModelBrand) => {
    // Redirect to version page details
    const brand = version.model?.brand?.name ?? "";
    const model = version.model?.name ?? "";
    const year = version.year.toString();
    const fullLabel = `${brand} ${model} ${year}`;
    setSearchQuery(fullLabel)
    setShowResults(false);
    router.push(`/versions/${version.id}`);
  }

  const handleInputFocus = () => {
    if (searchQuery.length > 0) {
      setShowResults(true);
    }
  }

  const handleKeyDown = (e: React.KeyboardEvent) => {
    if (e.key === "Escape") {
      setSearchQuery("");
      setShowResults(false);
      inputRef.current?.blur();
    }
  }

  useEffect(() => {
    function handleClickOutside(event: MouseEvent) {
      if (searchRef.current && !searchRef.current.contains(event.target as Node)) {
        setShowResults(false);
      }
    }

    document.addEventListener("mousedown", handleClickOutside);
    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, []);


  return (
    <div className="container flex h-16 items-center px-4">
      <div className="flex-1 max-w-md mx-4 relative" ref={searchRef}>
        <div className="relative">
          <Search className="absolute top-2.5 right-3 h-4 w-4 text-gray-500" />
          <Input
            ref={inputRef}
            type="text"
            placeholder="Buscar por marca, modelo o año..."
            value={searchQuery}
            onChange={(e) => setSearchQuery(e.target.value)}
            onFocus={handleInputFocus}
            onKeyDown={handleKeyDown}
            className="w-80 pr-10"
          />
        </div>

        {showResults && (
          // eslint-disable-next-line max-len
          <div className="absolute top-full left-0 right-0 mt-1 bg-background border rounded-md shadow-lg z-50 max-h-96 overflow-y-auto">
            {filteredVersions.length === 0 && searchQuery.length > 0 ? (
              <div className="p-4 text-center text-muted-foreground">No se encontraron versiones.</div>
            ) : (
              <div className="p-2">
                <div className="text-xs font-medium text-muted-foreground mb-2 px-2">Resultados de búsqueda</div>
                {filteredVersions.map((version) => {
                  const brand = version.model?.brand?.name ?? "";
                  const model = version.model?.name ?? "";
                  const year = version.year;
                  const label = `${brand} ${model} ${year}`;

                  return (
                    <div
                      key={version.id}
                      role="button"
                      tabIndex={0}
                      onClick={() => handleVersionSelect(version)}
                      onKeyDown={(e) => {
                        if (e.key === "Enter" || e.key === " ") {
                          handleVersionSelect(version);
                        }
                      }}
                      // eslint-disable-next-line max-len
                      className="flex items-center space-x-3 p-3 cursor-pointer hover:bg-accent rounded-md transition-colors"
                    >
                      <div className="flex-1">
                        <div className="font-medium">{label}</div>
                      </div>
                    </div>
                  )
                })}
              </div>
            )}
          </div>
        )}
      </div>
    </div>
  )
}
