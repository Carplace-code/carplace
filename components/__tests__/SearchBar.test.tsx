import { fireEvent, screen, waitFor } from "@testing-library/react";
import { vi } from "vitest";

import SearchBar from "@/components/SearchBar";
import * as useVersions from "@/hooks/useVersions";
import { renderWithQueryClient } from "@/utils/test-utils";

vi.mock("@/hooks/useVersions");

vi.mock("next/navigation", () => ({
  usePathname: () => "/",
  useRouter: () => ({
    push: vi.fn(),
    replace: vi.fn(),
    prefetch: vi.fn(),
    back: vi.fn(),
    forward: vi.fn(),
    refresh: vi.fn(),
  }),
}));

const mockData = {
  data: [
    {
      id: "1",
      year: 2020,
      model: {
        name: "Model S",
        brand: {
          name: "Tesla",
        },
      },
    },
    {
      id: "2",
      year: 2021,
      model: {
        name: "Model 3",
        brand: {
          name: "Tesla",
        },
      },
    },
    {
      id: "3",
      year: 2020,
      model: {
        name: "Civic",
        brand: {
          name: "Honda",
        },
      },
    },
  ],
};

describe("<SearchBar />", () => {
  beforeEach(() => {
    vi.clearAllMocks();
    vi.spyOn(useVersions, "useGetVersions").mockReturnValue({
      data: { data: mockData.data },
      isLoading: false,
      isError: false,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } as any);
  });

  it("renders the input correctly", () => {
    renderWithQueryClient(<SearchBar />);
    const input = screen.getByPlaceholderText("Buscar por marca, modelo o año...");
    expect(input).toBeInTheDocument();
  });

  it("shows filtered results when typing a brand", async () => {
    renderWithQueryClient(<SearchBar />);

    const input = screen.getByPlaceholderText("Buscar por marca, modelo o año...");

    fireEvent.change(input, { target: { value: "Tesla" } });

    await waitFor(() => {
      const results = screen.getAllByText(/Tesla/);
      expect(results.length).toBeGreaterThan(0);
    });
  });

  it("shows no results when the search does not match", async () => {
    renderWithQueryClient(<SearchBar />);

    const input = screen.getByPlaceholderText("Buscar por marca, modelo o año...");
    fireEvent.change(input, { target: { value: "XYZ123" } });

    await waitFor(() => {
      expect(screen.getByText("No se encontraron versiones.")).toBeInTheDocument();
    });
  });

  it("clears the results when input is emptied", async () => {
    renderWithQueryClient(<SearchBar />);

    const input = screen.getByPlaceholderText("Buscar por marca, modelo o año...");
    fireEvent.change(input, { target: { value: "Tesla" } });

    await waitFor(() => {
      const results = screen.getAllByText(/Tesla/);
      expect(results.length).toBeGreaterThan(0);
    });

    fireEvent.change(input, { target: { value: "" } });

    await waitFor(() => {
      expect(screen.queryByText(/Tesla/)).toBeNull();
    });
  });
  it("matches the snapshot", () => {
    const { container } = renderWithQueryClient(<SearchBar />);
    expect(container).toMatchSnapshot();
  });

  it("clears search on Escape key press", async () => {
    renderWithQueryClient(<SearchBar />);
    const input = screen.getByPlaceholderText("Buscar por marca, modelo o año...");
    fireEvent.change(input, { target: { value: "Civic" } });

    fireEvent.keyDown(input, { key: "Escape", code: "Escape" });

    await waitFor(() => {
      expect(input).toHaveValue("");
      expect(screen.queryByText(/Civic/)).toBeNull();
    });
  });

  it("shows results on input focus if searchQuery has value", async () => {
    renderWithQueryClient(<SearchBar />);
    const input = screen.getByPlaceholderText("Buscar por marca, modelo o año...");
    fireEvent.change(input, { target: { value: "Honda" } });

    fireEvent.focus(input);

    await waitFor(() => {
      expect(screen.getByText("Honda Civic 2020")).toBeInTheDocument();
    });
  });

  it("handles missing data gracefully", () => {
    vi.spyOn(useVersions, "useGetVersions").mockReturnValue({
      data: undefined,
      isLoading: false,
      isError: false,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } as any);

    renderWithQueryClient(<SearchBar />);
    const input = screen.getByPlaceholderText("Buscar por marca, modelo o año...");
    expect(input).toBeInTheDocument();
  });
});
