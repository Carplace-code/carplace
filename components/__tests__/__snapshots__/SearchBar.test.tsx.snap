// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<SearchBar /> > matches the snapshot 1`] = `
<div>
  <div
    class="container flex h-16 items-center px-4"
  >
    <div
      class="flex-1 max-w-md mx-4 relative"
    >
      <div
        class="relative"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-search absolute top-2.5 right-3 h-4 w-4 text-gray-500"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m21 21-4.34-4.34"
          />
          <circle
            cx="11"
            cy="11"
            r="8"
          />
        </svg>
        <input
          class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] disabled:opacity-50 md:text-sm aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive w-80 pr-10"
          data-slot="input"
          placeholder="Buscar por marca, modelo o aÃ±o..."
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;
